### ISSUE TITLE
Fault with Integration Chesterfield.gov.uk 

### ISSUE THREAD

**Issue by nathanael-pearson** [mampfes/hacs_waste_collection_schedule#386]
hi

This integration was working but is now showing the following error:Logger: waste_collection_schedule.scraper
I'm running the latest version of HA and the integration, I've tried rewriting the config and confirmed the UPRN is right.

--------------------------------------------------------------------------------------------------------------------------------------------------
Source: custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py:143

Integration: waste_collection_schedule ([documentation](https://github.com/mampfes/hacs_waste_collection_schedule#readme))
First occurred: 12:04:19 PM (2 occurrences)
Last logged: 1:16:23 PM

fetch failed for source chesterfield.gov.uk: Traceback (most recent call last): File
"/config/custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py", line 141, in fetch entries = self._source.fetch() File 
"/config/custom_components/waste_collection_schedule/waste_collection_schedule/source/chesterfield_gov_uk.py", line 74, in fetch waste_type = item["serviceTasks"][0]["taskTypeName"] IndexError: list index out of range
--------------------------------------------------------------------------------------------------------------------------------------------------

I've tried to run the debug test_sources.py but am getting access denied over SSH this script.

Kind regards

**Comment by dt215git** [comment]
The query now seems to also be returning an entry for Commercial waste collection schedules, but for a residential property it's empty. That's probably triggering the IndexError, and can be caught with _try except_ statement.
After a quick tinker, here's the test output having inserted the _try except_, you can see the "Collection entry contains no data" message being printed when the IndexError is caught.

```
Testing source chesterfield_gov_uk ...
Collection entry contains no data
  found 3 entries for Test_001
    2022-10-21: Domestic Recycling [mdi:recycle]
    2022-10-21: Domestic Organic [mdi:leaf]
    2022-10-14: Domestic Refuse [mdi:trash-can]
Collection entry contains no data
  found 3 entries for Test_002
    2022-10-19: Domestic Recycling [mdi:recycle]
    2022-10-19: Domestic Organic [mdi:leaf]
    2022-10-12: Domestic Refuse [mdi:trash-can]
Collection entry contains no data
  found 3 entries for Test_003
    2022-10-17: Domestic Refuse [mdi:trash-can]
    2022-10-10: Domestic Recycling [mdi:recycle]
    2022-10-10: Domestic Organic [mdi:leaf]
Collection entry contains no data
  found 3 entries for Test_004
    2022-10-18: Domestic Organic [mdi:leaf]
    2022-10-11: Domestic Refuse [mdi:trash-can]
    2022-10-18: Domestic Recycling [mdi:recycle]
```
I'll try and take a closer look this weekend.

**Comment by dt215git** [comment]
Update submitted: PR #391

**Comment by nathanael-pearson** [comment]
Hi @dt215git 

Thanks for the quick reply as always!

Kind regards

**Comment by nathanael-pearson** [comment]
Hi @dt215git 

Just checked the logs after the latest update and the error log has changed as below:

```
Logger: waste_collection_schedule.scraper
Source: custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py:143
Integration: waste_collection_schedule ([documentation](https://github.com/mampfes/hacs_waste_collection_schedule#readme))
First occurred: 8:04:57 PM (1 occurrences)
Last logged: 8:04:57 PM

fetch failed for source chesterfield.gov.uk: Traceback (most recent call last):
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py", line 141, in fetch entries = self._source.fetch()
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source/chesterfield_gov_uk.py", line 68, in fetch data = json.loads(r.content)
File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads return _default_decoder.decode(s)
File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/usr/local/lib/python3.10/json/decoder.py", line 355, in raw_decode raise JSONDecodeError("Expecting value", s, err.value) from None json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Is this a result of the last update or just a change in the response?
Kind regards

**Comment by dt215git** [comment]
The pull request hasn't been incorporated yet, so it's unlikely to be related to the change I made.
That error seems to suggest it didn't get anything back in response to the query. On the chesterfield_gov_uk.py file on your HA instance, you could try inserting a **print(r.content)** line immediately before the **data = json.loads(r.content)** line and seeing if anything appears in the logs after a restart.

**Comment by nathanael-pearson** [comment]
HI

Completed, log details below after a restart:

```
Logger: waste_collection_schedule.scraper
Source: custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py:143
Integration: waste_collection_schedule ([documentation](https://github.com/mampfes/hacs_waste_collection_schedule#readme))
First occurred: 10:44:22 PM (1 occurrences)
Last logged: 10:44:22 PM

fetch failed for source chesterfield.gov.uk: Traceback (most recent call last):
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py", line 141, in fetch entries = self._source.fetch()
File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source/chesterfield_gov_uk.py", line 70, in fetch data = json.loads(r.content)
File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads return _default_decoder.decode(s)
File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/usr/local/lib/python3.10/json/decoder.py", line 355, in raw_decode raise JSONDecodeError("Expecting value", s, err.value) from None json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Kind regards

**Comment by dt215git** [comment]
What do you get if you enter this url into a browser: [https://myaccount.chesterfield.gov.uk/anonymous/c/cbc_VE_CollectionDaysLO.app?aura.format=JSON&aura.formatAdapter=LIGHTNING_OUT](https://myaccount.chesterfield.gov.uk/anonymous/c/cbc_VE_CollectionDaysLO.app?aura.format=JSON&aura.formatAdapter=LIGHTNING_OUT)
It should return a load of json, but its the entries in `delegateVersion` and `auraConfig->context->fwuid` that might be most interesting.

**Comment by nathanael-pearson** [comment]
Hi

Thanks for updating the config for Chesterfield.
Since the latest version I now get the following...

`This error originated from a custom integration.

Logger: waste_collection_schedule.scraper
Source: custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py:143
Integration: waste_collection_schedule (documentation)
First occurred: 12:10:11 AM (1 occurrences)
Last logged: 12:10:11 AM

fetch failed for source chesterfield.gov.uk: Traceback (most recent call last): File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/scraper.py", line 141, in fetch entries = self._source.fetch() File "/config/custom_components/waste_collection_schedule/waste_collection_schedule/source/chesterfield_gov_uk.py", line 67, in fetch data = json.loads(r.content) File "/usr/local/lib/python3.10/json/__init__.py", line 346, in loads return _default_decoder.decode(s) File "/usr/local/lib/python3.10/json/decoder.py", line 337, in decode obj, end = self.raw_decode(s, idx=_w(s, 0).end()) File "/usr/local/lib/python3.10/json/decoder.py", line 355, in raw_decode raise JSONDecodeError("Expecting value", s, err.value) from None json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`

I tried to track back the guid you referenced but I think it's a little above my skills to trace the source of the fault.

**Comment by dt215git** [comment]
That's still suggesting it's not getting a response to the query.

If you enter [https://myaccount.chesterfield.gov.uk/anonymous/c/cbc_VE_CollectionDaysLO.app?aura.format=JSON&aura.formatAdapter=LIGHTNING_OUT](https://myaccount.chesterfield.gov.uk/anonymous/c/cbc_VE_CollectionDaysLO.app?aura.format=JSON&aura.formatAdapter=LIGHTNING_OUT) into a browser, you should see something like the Firefox screenshot below. I've highlighted `delegateVersion` and `auraConfig->context->fwuid`. I'm curious what value you see for those.

![image](https://user-images.githubusercontent.com/86194065/197327185-cb6a546b-ca28-4649-bb8f-d65c1baa718a.png)

**Comment by nathanael-pearson** [comment]
Hi

I get the following:

![image](https://user-images.githubusercontent.com/69221775/197339284-74b8e747-31e5-41ca-ae2c-79b50c784bcb.png)

They both show:  "izKqHkAqhQoblYy3lGicEA"

Kind regards

**Comment by dt215git** [comment]
Thanks.
I think the fwuid is periodically changing and that breaks the query - that would explain why it originally worked for you and then suddenly stopped, and why changes I make locally are failing by the time they get pushed out in a new release.

If you look in your HA chesterfield_gov_uk.py file you should see this code:
```python
payload = {
    "message": '{\"actions\":[{\"id\":\"4;a\",\"descriptor\":\"aura://ApexActionController/ACTION$execute\",\"callingDescriptor\":\"UNKNOWN\",\"params\":{\"namespace\":\"\",\"classname\":\"CBC_VE_CollectionDays\",\"method\":\"getServicesByUPRN\",\"params\":{\"propertyUprn\":\"' + self._uprn + '\",\"executedFrom\":\"Main Website\"},\"cacheable\":false,\"isContinuation\":false}}]}',
    "aura.context": '{\"mode\":\"PROD\",\"fwuid\":\"5FtqNRNwJDpZNZFKfXyAmg\",\"app\":\"c:cbc_VE_CollectionDaysLO\",\"loaded\":{\"APPLICATION@markup://c:cbc_VE_CollectionDaysLO\":\"pqeNg7kPWCbx1pO8sIjdLA\"},\"dn\":[],\"globals\":{},\"uad\":true}',
    "aura.pageURI": '/bins-and-recycling/bin-collections/check-bin-collections.aspx',
    "aura.token": 'null'
}
```

You can see `\"fwuid\":\"5FtqNRNwJDpZNZFKfXyAmg\"` on the _aura.context_ line. If you replace _5FtqNRNwJDpZNZFKfXyAmg_ with _zKqHkAqhQoblYy3lGicEA_ and restart HA it should hopefully work (providing it hasn't changed again).

The PR(#402 ) I submitted a short while go tries to deal with a changing fwuid by capturing it from the url we've just used, and then substituting it into the later query for the collection schedule.

If that PR doesn't work, then I'm out of ideas!

**Comment by mampfes** [comment]
fixed in #402
